experimental/dds/tree2/src/feature-libraries/nodeIdentifierIndex.ts

```diff
@@ -120,50 +74,50 @@ export class IdentifierIndex<TField extends GlobalFieldKey>
 	public forEach(
 		callbackfn: (
 			value: EditableTree,
-			key: Identifier,
-			map: ReadonlyMap<Identifier, EditableTree>,
+			key: NodeIdentifier,
+			map: ReadonlyMap<NodeIdentifier, EditableTree>,
 		) => void,
 		thisArg?: any,
 	): void {
 		return this.nodes.forEach(callbackfn, thisArg);
 	}
-	public get(key: Identifier): EditableTree | undefined {
+	public get(key: NodeIdentifier): EditableTree | undefined {
 		return this.nodes.get(key);
 	}
-	public has(key: Identifier): boolean {
+	public has(key: NodeIdentifier): boolean {
 		return this.nodes.has(key);
 	}
 	public get size(): number {
 		return this.nodes.size;
 	}
-	public entries(): IterableIterator<[Identifier, EditableTree]> {
+	public entries(): IterableIterator<[NodeIdentifier, EditableTree]> {
 		return this.nodes.entries();
 	}
-	public keys(): IterableIterator<Identifier> {
+	public keys(): IterableIterator<NodeIdentifier> {
 		return this.nodes.keys();
 	}
 	public values(): IterableIterator<EditableTree> {
 		return this.nodes.values();
 	}
-	public [Symbol.iterator](): IterableIterator<[Identifier, EditableTree]> {
+	public [Symbol.iterator](): IterableIterator<[NodeIdentifier, EditableTree]> {
 		return this.nodes[Symbol.iterator]();
 	}
 	// #endregion ReadonlyMap interface
 
 	private *findIdentifiers(
 		node: EditableTree,
-	): Iterable<[identifier: Identifier, node: EditableTree]> {
+	): Iterable<[identifier: NodeIdentifier, node: EditableTree]> {
 		if (this.identifierFieldKeySymbol in node) {
 			const type = node[typeSymbol];
 			if (type.extraGlobalFields || type.globalFields.has(this.identifierFieldKey)) {
 				// Get the ID via a wrapped node rather than an unwrapped node (`node[identifierFieldKeySymbol]`)
-				// so that the length of the field and the type of the value can be validated as correct
+				// so that the field kind can be checked
 				const field = node[getField](this.identifierFieldKeySymbol);
-				if (field.length >= 1) {
+				if (field.fieldSchema.kind.identifier === nodeIdentifier.identifier) {
 					const identifierNode = field.getNode(0);
 					const id = identifierNode[valueSymbol];
-					if (isIdentifier(id)) {
-						yield [id, node];
+					if (typeof id === "string" && isStableId(id)) {
```

Comment:
findIdentifiers should only run if you have already determined the document has the expected schema for the global field right?

Therefor if the value on the node is not a string, the data is out of schema and this should probably assert.

If isStableId fails, thats also an error, and should probably also assert.

Reply:
You're right, I think we should fail too. I added an assert and updated the tests.